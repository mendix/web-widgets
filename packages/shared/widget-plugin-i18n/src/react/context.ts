import { createContext } from "react";
import { PluginI18n } from "../types";

const contextPath = "com.mendix.widget.plugin.i18n.context" as const;

type I18nContext = React.Context<PluginI18n | null>;

const createI18nContext = (): I18nContext => createContext<PluginI18n | null>(null);

declare global {
    interface Window {
        [contextPath]?: ReturnType<typeof createI18nContext>;
    }
}

export function getGlobalI18nPluginContext(): I18nContext {
    return window[contextPath] ?? createI18nContext();
}
