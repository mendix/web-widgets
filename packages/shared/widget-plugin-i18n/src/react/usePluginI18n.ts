import { useState } from "react";
import { I18nController } from "../controllers/I18nController";
import { PluginI18n } from "../types";

export function usePluginI18n(languageCode: string): PluginI18n | null {
    const [plugin, setPlugin] = useState<PluginI18n | null>(() => {
        new I18nController({
            languageCode,
            onPluginChange: plugin => setPlugin(plugin)
        });

        return null;
    });

    return plugin;
}
