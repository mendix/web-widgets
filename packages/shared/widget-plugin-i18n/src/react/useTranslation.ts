import { useContext } from "react";
import { PluginI18n } from "../types";
import { getGlobalI18nPluginContext } from "./context";

export function useTranslation(): PluginI18n | null {
    return useContext(getGlobalI18nPluginContext());
}
