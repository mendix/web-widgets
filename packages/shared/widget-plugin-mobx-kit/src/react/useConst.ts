import { useRef } from "react";

export function useConst<T>(fn: T): T extends (...args: never) => infer R ? R : T {
    return (useRef<T | null>(null).current ??= typeof fn === "function" ? fn() : fn);
}
