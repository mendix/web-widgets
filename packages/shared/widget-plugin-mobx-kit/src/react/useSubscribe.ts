import { IReactionOptions, reaction } from "mobx";
import { useEffect, useState } from "react";

export function useSubscribe<Store extends object, T, FireImmediately extends boolean = false>(
    init: Store | (() => Store),
    selector: (store: Store) => T,
    options?: IReactionOptions<T, FireImmediately>
): [T, Store] {
    const [store] = useState(() => (typeof init === "function" ? init() : init));
    const [value, setValue] = useState(() => selector(store));

    function setup(): () => void {
        return reaction(() => selector(store), setValue, options);
    }

    // eslint-disable-next-line react-hooks/exhaustive-deps
    useEffect(setup, []);

    return [value, store];
}
