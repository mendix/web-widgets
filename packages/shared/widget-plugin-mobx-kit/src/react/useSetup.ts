import { useEffect } from "react";
import { useConst } from "./useConst";

export function useSetup<
    T extends {
        setup(): void | (() => void);
    }
>(fn: () => T): T {
    const obj = useConst(fn);
    useEffect(() => obj.setup(), [obj]);
    return obj;
}
