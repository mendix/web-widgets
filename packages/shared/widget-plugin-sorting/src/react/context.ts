import { useConst } from "@mendix/widget-plugin-mobx-kit/react/useConst";
import { generateUUID } from "@mendix/widget-plugin-platform/framework/generate-uuid";
import { Context, createContext, useContext, useEffect } from "react";
import { SortStoreHost } from "../stores/SortStoreHost";
import { Result, error, value } from "./result-meta";

export interface SortAPI {
    version: 1;
    host: SortStoreHost;
}

const SORT_PATH = "com.mendix.widgets.web.sortable.sortContext";

export function getGlobalSortContext(): Context<SortAPI | null> {
    const scope = window.top === window ? window : window.top;
    return ((scope as any)[SORT_PATH] ??= createContext<SortAPI | null>(null));
}

export function useSortAPI(): Result<SortAPI, Error> {
    const api = useContext(getGlobalSortContext());
    if (api === null) {
        return error(new Error("Error: widget is out of context. Please place the widget inside the Gallery header."));
    }
    return value(api);
}

export function useLockSortAPI(api: SortAPI): Result<SortAPI, Error> {
    const id = useLock(api);

    if (api.host.usedBy !== id) {
        return error(
            new Error(
                `Error: Sort API is already in use by another widget. Remove other sort widgets and refresh the page.`
            )
        );
    }

    return value(api);
}

function useLock({ host }: SortAPI): string {
    const [unlock, id] = useConst(() => {
        const id = `useLock@${generateUUID()}`;
        return [host.lock(id), id] as const;
    });

    useEffect(() => unlock, [unlock]);

    return id;
}
