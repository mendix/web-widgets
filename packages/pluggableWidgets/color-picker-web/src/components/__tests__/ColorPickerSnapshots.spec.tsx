import "@testing-library/jest-dom";
import { act, render, RenderResult } from "@testing-library/react";
import { ColorPicker, ColorPickerProps } from "../ColorPicker";

jest.mock("react-color/lib", () => ({
    SketchPicker: "sketch-picker"
}));

describe("ColorPicker", () => {
    let colorPickerProps: ColorPickerProps;

    beforeEach(() => {
        colorPickerProps = {
            color: "#000000",
            disabled: false,
            defaultColors: [],
            format: "hex",
            mode: "popover",
            type: "sketch",
            onChange: jest.fn(),
            onColorChange: jest.fn(),
            id: "color-picker",
            name: "color picker"
        };
    });

    function renderColorPicker(configs: Partial<ColorPickerProps> = {}): RenderResult {
        return render(<ColorPicker {...colorPickerProps} {...configs} />);
    }

    it("renders the structure correctly", () => {
        const colorPickerComponent = renderColorPicker();

        expect(colorPickerComponent.asFragment()).toMatchSnapshot();
    });

    it("that is disabled renders with the structure", () => {
        const colorPickerComponent = renderColorPicker({ disabled: true });
        expect(colorPickerComponent.asFragment()).toMatchSnapshot();
    });

    it("renders picker with pre-defined default colors", () => {
        const colorPickerComponent = renderColorPicker();

        act(() => {
            colorPickerComponent.rerender(
                <ColorPicker
                    {...colorPickerProps}
                    mode="inline"
                    defaultColors={[{ color: "#2CCCE4" }, { color: "#555555" }]}
                />
            );
        });

        expect(colorPickerComponent.asFragment()).toMatchSnapshot();
    });

    describe("with a mode as", () => {
        it("popover or input renders with the structure", () => {
            const colorPickerComponent = renderColorPicker({ mode: "popover" });

            expect(colorPickerComponent.asFragment()).toMatchSnapshot();
        });

        it("inline renders with the structure", () => {
            const colorPickerComponent = renderColorPicker({ mode: "inline" });

            expect(colorPickerComponent.asFragment()).toMatchSnapshot();
        });
    });
});
