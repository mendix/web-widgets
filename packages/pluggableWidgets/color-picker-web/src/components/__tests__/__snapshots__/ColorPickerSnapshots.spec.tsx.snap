// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ColorPicker renders picker with pre-defined default colors 1`] = `
<DocumentFragment>
  <div
    class="widget-color-picker widget-color-picker-picker"
  />
</DocumentFragment>
`;

exports[`ColorPicker renders the structure correctly 1`] = `
<DocumentFragment>
  <div
    class="widget-color-picker widget-color-picker-picker"
  >
    <button
      class="btn"
    >
      <div
        class="widget-color-picker-inner"
        style="background: rgb(0, 0, 0);"
      />
    </button>
  </div>
</DocumentFragment>
`;

exports[`ColorPicker that is disabled renders with the structure 1`] = `
<DocumentFragment>
  <div
    class="widget-color-picker widget-color-picker-picker widget-color-picker-disabled"
  >
    <button
      class="btn disabled"
    >
      <div
        class="widget-color-picker-inner"
        style="background: rgb(0, 0, 0);"
      />
    </button>
  </div>
</DocumentFragment>
`;

exports[`ColorPicker with a mode as inline renders with the structure 1`] = `
<DocumentFragment>
  <div
    class="widget-color-picker widget-color-picker-picker"
  >
    <div
      class=""
    >
      <sketch-picker
        color="#000000"
        disablealpha="true"
        disabled="false"
        displaycolorpicker="false"
      />
    </div>
  </div>
</DocumentFragment>
`;

exports[`ColorPicker with a mode as popover or input renders with the structure 1`] = `
<DocumentFragment>
  <div
    class="widget-color-picker widget-color-picker-picker"
  >
    <button
      class="btn"
    >
      <div
        class="widget-color-picker-inner"
        style="background: rgb(0, 0, 0);"
      />
    </button>
  </div>
</DocumentFragment>
`;
